modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/impl/TaskTriggerHandler.java

```diff
@@ -169,12 +172,25 @@ private void handleTask(Task task, Map<String, Object> parameters, boolean isRet
 
         TaskContext taskContext = new TaskContext(task, parameters, metadata, activityService);
         TaskInitializer initializer = new TaskInitializer(taskContext);
+        List<FilterSet> filterSet = new ArrayList<>(task.getTaskConfig().getFilters());
+        TaskFilterExecutor taskFilterExecutor = new TaskFilterExecutor();
+        boolean result = true;
+        int executedActions = 0;
+        int checkedFilters = 0;
+        int actualFilterIndex = initializer.getActionFilters();
 
         try {
             LOGGER.info("Executing all actions from task: {}", task.getName());
             if (initializer.evalConfigSteps(dataProviders)) {
-                for (int i = 0; i < task.getActions().size(); i++) {
-                    executor.execute(task, task.getActions().get(i), i, taskContext, activityId);
+                while (result && executedActions < task.getActions().size()) {
```

Comment:
the same, it is difficult to understand it. Try to move sth into new method and give proper name